pkg/storage/etcd3/watcher.go

```diff
@@ -322,7 +322,13 @@ func prepareObjs(ctx context.Context, e *event, client *clientv3.Client, codec r
 		if err != nil {
 			return nil, nil, err
 		}
-		oldObj, err = decodeObj(codec, versioner, getResp.Kvs[0].Value, getResp.Kvs[0].ModRevision)
+		rev := getResp.Kvs[0].ModRevision
+		if e.isDeleted {
+			// Note that this sends the *old* object with the etcd revision for the time at
+			// which it gets deleted. Users use that revision to tell ordering.
+			rev = e.rev
+		}
+		oldObj, err = decodeObj(codec, versioner, getResp.Kvs[0].Value, rev)
```

Comment:
They use the event number to signal the last known resource version in the informer in case there is a watch break.